gdb_modules/gala_compatibility.py

```diff
@@ -0,0 +1,43 @@
+"""GALA compatibility classes.
+
+These classes expose lldb functionality that doesn't have an exact equivalent in
+the gdb API, in a way that's compatible with both debuggers.
+"""
+import gdb
+import gdb.printing
+
+class TypeCallbackPrettyPrinter(gdb.printing.PrettyPrinter):
+  """A type-callback prettyprinter compatible with GDB and lldb-with-GALA."""
+
+  # The Subprinter class doesn't have any logic. We need it here because GALA's
+  # implementation of `gdb.register_pretty_printer()` uses the
+  # `gala_matching_function` and `gala_make_printer_function` subprinter
+  # attributes to drive callback-based formatter registration in LLDB.
+  class Subprinter:
```

Comment:
Sounds like this could just be a data class

```
import dataclasses
@dataclasses.dataclass
  class Subprinter:
    enabled: bool=True
    name: str
    gala_matching_function: Any # or Callable[[gdb.Type], bool]?
    gala_make_printer_function: Any # or Callable[[gdb.Value], gdb.printing.PrettyPrinter]?
```

Is there a reason it's an inner class? I don't think inner classes have any advantage in Python, e.g. they don't have special scope access. I wouldn't expect to see an inner class unless it's something that's super tightly coupled with the outer class, while this seems only medium-ly coupled.

Reply:
dataclasses were introduced in Python 3.7 IIUC, and this module should be compatible with Python 2 for now.

The reason it's an inner class is mainly namespacing. The subprinter is an implementation detail and other classes can define their own subprinter class. It's also a class that won't ever be instantiated from outside of the main prettyprinter class, so it doesn't make a lot of sense for it to be module-level (maybe we could name it `_Subprinter` to make it even clearer?). I guess what I'm trying to say is that they do seem tightly coupled to me. If I were to move it outside the main class I'd name it something like `TypeCallbackSubPrettyPrinter` and I'm not sure that's much of a win.

But I don't have a very strong opinion about this. If you do, I can move it out.

BTW I just noticed that this should inherit from `gdb.printing.SubPrettyPrinter`, mostly for self-documenting reasons (gdb doesn't actually care, as long as it has all required members). I'll make that change tomorrow.

Reply:
SGTM, leaving it as an inner class is fine then.